cpp/include/patterns/copy_v_transform_reduce_nbr.cuh

```diff
@@ -0,0 +1,293 @@
+                   vertex_value_output_first,
+                   vertex_value_output_first + graph_view.get_number_of_local_vertices(),
+                   init);
+    }
+
+    assert(graph_view.get_number_of_local_vertices() ==
+           graph_view.get_number_of_adj_matrix_local_rows());
+    assert(graph_view.get_number_of_local_vertices() ==
+           graph_view.get_number_of_adj_matrix_local_cols());
+    detail::for_all_major_for_all_nbr_low_out_degree<GraphViewType::is_adj_matrix_transposed>
+      <<<update_grid.num_blocks, update_grid.block_size, 0, handle.get_stream()>>>(
+        matrix_partition,
+        adj_matrix_row_value_input_first,
+        adj_matrix_col_value_input_first,
+        vertex_value_output_first,
+        e_op,
+        init);
+  }
+}
+
+/**
+ * @brief Iterate over the outgoing edges to update vertex properties.
+ *
+ * This function is inspired by thrust::transfrom_reduce() (iteration over the outgoing edges part)
+ * and thrust::copy() (update vertex properties part, take transform_reduce output as copy input).
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @tparam AdjMatrixRowValueInputIterator Type of the iterator for graph adjacency matrix row
+ * input properties.
+ * @tparam AdjMatrixColValueInputIterator Type of the iterator for graph adjacency matrix column
+ * input properties.
+ * @tparam EdgeOp Type of the binary (or ternary) edge operator.
+ * @tparam T Type of the initial value for reduction over the outgoing edges.
+ * @tparam VertexValueOutputIterator Type of the iterator for vertex output property variables.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param adj_matrix_row_value_input_first Iterator pointing to the adjacency matrix row input
+ * properties for the first (inclusive) row (assigned to this process in multi-GPU).
+ * `adj_matrix_row_value_input_last` (exclusive) is deduced as @p adj_matrix_row_value_input_first +
+ * @p graph_view.get_number_of_adj_matrix_local_rows().
+ * @param adj_matrix_col_value_input_first Iterator pointing to the adjacency matrix column input
+ * properties for the first (inclusive) column (assigned to this process in multi-GPU).
+ * `adj_matrix_col_value_output_last` (exclusive) is deduced as @p adj_matrix_col_value_output_first
+ * + @p graph_view.get_number_of_adj_matrix_local_cols().
+ * @param e_op Binary (or ternary) operator takes *(@p adj_matrix_row_value_input_first + i), *(@p
+ * adj_matrix_col_value_input_first + j), (and optionally edge weight) (where i and j are row and
+ * column indices, respectively) and returns a value to be reduced.
+ * @param init Initial value to be added to the reduced @e_op return values for each vertex.
+ * @param vertex_value_output_first Iterator pointing to the vertex property variables for the first
```

Comment:
I was confused by this while trying to use this function.  If I pass `output_v.begin()` this fails to compile, I need to pass `output_v.data().get()`.

It would be helpful to either be able to pass a thrust iterator (`thrust::copy_if` takes an iterator) or to document that you need to pass a raw pointer (or whatever the actual criteria is).

Reply:
This sounds like a bug, this should work with both .begin() and .data().get(). I will take a look.

Reply:
This should be fixed by now. vertex_value_output_first can be either raw pointer, thrust::device_ptr, thrust::discard_iterator, or thrust::zip_iterator having a mix of raw pointer, thrust::device_ptr, and/or thrust::discard_iterator. Actual values pointed by either raw pointer or thrust::device_ptr should be arithmetic type.